waku/waku_rln_relay/group_manager/on_chain/group_manager.nim

```diff
@@ -616,25 +618,42 @@ proc startOnchainSync(
         blockNumber = g.rlnContractDeployedBlockNumber
       g.rlnContractDeployedBlockNumber
 
+  var futs = newSeq[Future[bool]]()
+  var currentLatestBlock: BlockNumber
+  g.retryWrapper(currentLatestBlock, "Failed to get the latest block number"):
+    cast[BlockNumber](await ethRpc.provider.eth_blockNumber())
+  
   try:
     # we always want to sync from last processed block => latest
     # chunk events
     while true:
-      var currentLatestBlock: BlockNumber
-      g.retryWrapper(currentLatestBlock, "Failed to get the latest block number"):
-        cast[BlockNumber](await ethRpc.provider.eth_blockNumber())
+      # if the fromBlock is less than 2k blocks behind the current block
+      # then fetch the new toBlock
       if fromBlock >= currentLatestBlock:
         break
+      
+      if fromBlock + blockChunkSize.uint > currentLatestBlock.uint:
+        g.retryWrapper(currentLatestBlock, "Failed to get the latest block number"):
+          cast[BlockNumber](await ethRpc.provider.eth_blockNumber())
+      
 
       let toBlock = min(fromBlock + BlockNumber(blockChunkSize), currentLatestBlock)
       debug "fetching events", fromBlock = fromBlock, toBlock = toBlock
-      var handleBlockRes: bool
-      g.retryWrapper(handleBlockRes, "Failed to handle old blocks"):
-        await g.getAndHandleEvents(fromBlock, toBlock)
+      await sleepAsync(rpcDelay)
+      futs.add(g.getAndHandleEvents(fromBlock, toBlock))
       fromBlock = toBlock + 1
+    for fut in futs:
```

Comment:
am i missing something or is this unbounded? i mean what if the sync distance is 2 million blocks? will we spawn 2mill/2000 rpc calls?
If so we may:
* hit some OS limits? unsure if you can spawn async eg 200 calls such this?
* kill the poor eth node? specially if someone is using its own (which will most likely have limited resources).

Reply:
addressed in [7116bff](https://github.com/waku-org/nwaku/pull/2577/commits/7116bff97b70603feb0b5967fb56c4546a86dee5), moderately affects sync time, it is now ~3 mins on the same connection i tested previously with.